packages/data-source/src/DataSourceWidget.tsx

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2018-2020 IBM Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import '../style/index.css';
+
+import { ExpandableComponent, insertCodeSnippet } from '@elyra/ui-components';
+
+import { JupyterFrontEnd } from '@jupyterlab/application';
+import { ReactWidget, UseSignal } from '@jupyterlab/apputils';
+import { IDocumentManager } from '@jupyterlab/docmanager';
+import { IRenderMimeRegistry } from '@jupyterlab/rendermime';
+import { addIcon } from '@jupyterlab/ui-components';
+
+import { CommandRegistry } from '@lumino/commands';
+import { Message } from '@lumino/messaging';
+import { Signal } from '@lumino/signaling';
+import { Widget, Menu } from '@lumino/widgets';
+
+import React from 'react';
+
+import { CodeTemplateManager } from './CodeTemplate';
+import { DataSourceManager, IDataSource } from './DataSource';
+
+/**
+ * The CSS class added to data source widget.
+ */
+const DATA_SOURCES_CLASS = 'elyra-DataSources';
+const DATA_SOURCES_HEADER_CLASS = 'elyra-dataSourcesHeader';
+const DATA_SOURCE_ITEM = 'elyra-dataSource-item';
+
+/**
+ * DataSourceDisplay props.
+ */
+interface IDataSourceDisplayProps {
+  dataSources: IDataSource[];
+  getCurrentWidget: () => Widget;
+  app: JupyterFrontEnd;
+}
+
+/**
+ * A React Component for data-sources display list.
+ */
+class DataSourceDisplay extends React.Component<IDataSourceDisplayProps> {
+  props: any;
+  commandRegistry: CommandRegistry;
+
+  constructor(props: any) {
+    super(props);
+  }
+
+  private addSubmenu(dataSource: IDataSource): void {
+    const languageMenus: any[] = [];
+    const id = dataSource.id;
+    for (const template of this.props.codeTemplates) {
+      const language = template.language;
+      const framework = template.framework;
+
+      // Only show code templates that are for the format of this datasource
+      const format = template.format;
+      if (format != dataSource.format) {
+        continue;
+      }
+      const menuObj: any = languageMenus.find((languageMenu: any) => {
+        return languageMenu.language == language;
+      });
+      let menu: Menu;
+      if (!menuObj) {
+        menu = new Menu({ commands: this.commandRegistry });
```

Comment:
For some reason, every time I click on some other extension button and come back to data-sources extension then right-click the `+` button, a new submenu is created. So I get something like 
![image](https://user-images.githubusercontent.com/25207344/83077840-4ab8c780-a046-11ea-9fcd-a4c37872888e.png)
Looks like `languageMenus` array is being cleared every time render() is called. Perhaps `addSubMenu` function can be called from some other event like on `ComponentDidMount` or something?

Also, is the default msg in context menu necessary/applies here "Shift+Right click for Browser Menu"? Not sure it can be overridden..